testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java

```diff
@@ -52,14 +52,146 @@
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.junit.jupiter.api.Assumptions.assumeFalse;
 
+@Timeout(value = 20000, unit = MILLISECONDS)
 public class SocketHalfClosedTest extends AbstractSocketTest {
 
     protected int maxReadCompleteWithNoDataAfterInputShutdown() {
         return 2; // nio needs read flag to detect full closure.
     }
 
     @Test
-    @Timeout(value = 5000, unit = MILLISECONDS)
+    public void testAllDataReadEventTriggeredAfterHalfClosure(TestInfo testInfo) throws Throwable {
+        run(testInfo, new Runner<ServerBootstrap, Bootstrap>() {
+            @Override
+            public void run(ServerBootstrap serverBootstrap, Bootstrap bootstrap) throws Throwable {
+                // TODO: there needs to be a more robust way to detect this
+                if (bootstrap.toString().contains("NioSocketChannel") ||
+                        bootstrap.toString().contains("OioSocketChannel")) {
+                    return;
+                }
+                allDataReadEventTriggeredAfterHalfClosure(serverBootstrap, bootstrap);
+            }
+        });
+    }
+
+    private void allDataReadEventTriggeredAfterHalfClosure(ServerBootstrap sb, Bootstrap cb) throws Throwable {
+        final int totalServerBytesWritten = 1;
+        final CountDownLatch clientReadAllDataLatch = new CountDownLatch(1);
+        final CountDownLatch clientHalfClosedLatch = new CountDownLatch(1);
+        final CountDownLatch clientHalfClosedAllBytesRead = new CountDownLatch(1);
+        final AtomicInteger clientReadCompletes = new AtomicInteger();
+        final AtomicInteger clientZeroDataReadCompletes = new AtomicInteger();
+        Channel serverChannel = null;
+        Channel clientChannel = null;
+        AtomicReference<Channel> serverChildChannel = new AtomicReference<>();
+        try {
+            cb.option(ChannelOption.ALLOW_HALF_CLOSURE, true)
+                    .option(ChannelOption.AUTO_CLOSE, false)
+                    .option(ChannelOption.AUTO_READ, false);
+
+            sb.childHandler(new ChannelInitializer<Channel>() {
```

Comment:
Since we're only sending a single byte from the server, we can make the test run a little bit faster with `sb.childOption(ChannelOption.TCP_NODELAY, true);`